hls-test-utils/src/Test/Hls/FileSystem.hs

```diff
@@ -99,15 +101,23 @@ materialise rootDir' fileTree testDataDir' = do
       rootDir = FP.normalise rootDir'
 
       persist :: FilePath -> FileTree -> IO ()
-      persist fp (File name cts) = case cts of
-        Inline txt -> T.writeFile (fp </> name) txt
-        Ref path -> copyFile (testDataDir </> FP.normalise path) (fp </> takeFileName name)
-      persist fp (Directory name nodes) = do
-        createDirectory (fp </> name)
-        mapM_ (persist (fp </> name)) nodes
+      persist root (File name cts) = case cts of
+        Inline txt -> T.writeFile (root </> name) txt
+        Ref path -> copyFile (testDataDir </> FP.normalise path) (root </> takeFileName name)
+      persist root (Directory name nodes) = do
+        createDirectory (root </> name)
+        mapM_ (persist (root </> name)) nodes
+      persist root (CopiedDirectory name) = do
+        nodes <- copyDir' testDataDir' name
+        mapM_ (persist root) nodes
 
   traverse_ (persist rootDir) fileTree
   pure $ FileSystem rootDir fileTree testDataDir
+    where -- | Copy a directory into a test project.
+        copyDir' :: FilePath -> FilePath -> IO [FileTree]
+        copyDir' root dir = do
+            files <- listDirectory (root </> dir)
+            traverse (\f -> pure $ copy (dir </> f)) files
```

Comment:
I think we should only copy files that have been added to the git index. Cabal testsuite does something similar: https://github.com/haskell/cabal/blob/master/cabal-testsuite/src/Test/Cabal/Monad.hs#L417

Reply:
Great, let's copy some of the cabal code

Reply:
One concern though, does it mean we have to stage the files if newly test files are added.

Reply:
Right the `--modified` should list unstaged files